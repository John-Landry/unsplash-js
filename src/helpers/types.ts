// Copied from https://github.com/Microsoft/TypeScript/issues/1897#issuecomment-338650717
export type AnyJson = boolean | number | string | null | JsonArray | JsonMap;
export type JsonMap = { [key: string]: AnyJson };
export type JsonArray = Array<AnyJson>;

// https://stackoverflow.com/a/9436948
export const checkIsString = (value: unknown): value is string =>
  typeof value === 'string' || value instanceof String;

// https://github.com/Microsoft/TypeScript/issues/12215#issuecomment-377567046
export type OmitStrict<T, K extends keyof T> = Omit<T, K>;

// Note: unlike TypeScript's `NonNullable`, this does _not_ include `undefined`. See discussion at
export type Nullable<T> = T | null;

const isDefined = <T>(x: T | null | undefined): x is T =>
  x !== null && x !== undefined;

/** Takes a dictionary containing nullish values and returns a dictionary of all the defined
 * (non-nullish) values.
 *
 * Inspired by
 * http://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Maybe.html#v:catMaybes
 * https://github.com/gcanti/fp-ts/blob/42870714ebcae4ecf132291c687c845b6837b7d2/docs/Record.md#compact
 * Similar to fp-ts `compact`, but uses `Defined` instead of `Option`.
 */
export const compactDefined = <A>(obj: Record<string, A>) =>
  Object.keys(obj).reduce<Record<string, A>>((acc, key) => {
    const value = obj[key];
    return {
      ...acc,
      ...(isDefined(value) ? { key: value } : {}),
    };
  }, {});

export enum OrderBy {
  LATEST = 'latest',
  POPULAR = 'popular',
  OLDEST = 'oldest',
}
export type Orientation = 'landscape' | 'portrait' | 'squarish';
