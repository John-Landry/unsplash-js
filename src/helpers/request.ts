import { ParsedUrlQueryInput } from 'querystring';
import { addQueryToUrl, appendPathnameToUrl } from 'url-transformers';
import { compactDefined, flow } from './fp';
import { ApiResponse, handleFetchResponse, HandleResponse } from './response';
import { isDefined, OmitStrict } from './typescript';

type BuildUrlParams = {
  pathname: string;
  query?: ParsedUrlQueryInput;
};

const buildUrl = ({ pathname, query = {} }: BuildUrlParams) =>
  flow(appendPathnameToUrl(pathname), addQueryToUrl(compactDefined(query)));

type FetchParams = Pick<RequestInit, 'method'>;
/**
 * The params generated by the library
 */
type BaseRequestParams = BuildUrlParams & FetchParams;

/**
 * Additional fetch options provided by the user on a per-call basis
 */
type AdditionalPerFetchParams = Omit<RequestInit, keyof FetchParams>;
type CompleteRequestParams = BaseRequestParams & AdditionalPerFetchParams;

/**
 * helper used to type-check the arguments, and add default params for all requests
 */
export const createRequestHandler = <Args extends {}>(fn: (a: Args) => BaseRequestParams) => (
  a: Args,
  additionalFetchOptions: AdditionalPerFetchParams = {},
): CompleteRequestParams => ({
  ...fn(a),
  ...additionalFetchOptions,
});

/**
 * Initial parameters that apply to all calls
 */
type InitParams = {
  accessKey?: string;
  apiVersion?: string;
  apiUrl?: string;
} & OmitStrict<RequestInit, 'method' | 'body'>;

type RequestGenerator<RequestArgs extends unknown[], ResponseType> = {
  handleRequest: (...a: RequestArgs) => CompleteRequestParams;
  handleResponse: HandleResponse<ResponseType>;
};

type InitMakeRequest = (
  args: InitParams,
) => <RequestArgs extends unknown[], ResponseType>(
  handlers: RequestGenerator<RequestArgs, ResponseType>,
) => (...a: Parameters<typeof handlers['handleRequest']>) => Promise<ApiResponse<ResponseType>>;

export const initMakeRequest: InitMakeRequest = ({
  accessKey,
  apiVersion = 'v1',
  apiUrl = 'https://api.unsplash.com',
  headers: generalHeaders,
  ...generalFetchOptions
}) => ({ handleResponse, handleRequest }) =>
  flow(
    handleRequest,
    ({ pathname, query, method = 'GET', headers: endpointHeaders, body, signal }) => {
      const url = buildUrl({ pathname, query })(apiUrl);

      const fetchOptions: RequestInit = {
        method,
        headers: {
          ...generalHeaders,
          ...endpointHeaders,
          'Accept-Version': apiVersion,
          ...(isDefined(accessKey) ? { Authorization: `Client-ID ${accessKey}` } : {}),
        },
        body,
        signal,
        ...generalFetchOptions,
      };

      return fetch(url, fetchOptions).then(handleFetchResponse(handleResponse));
    },
  );
