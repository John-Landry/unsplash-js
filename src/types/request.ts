export enum OrderBy {
  LATEST = 'latest',
  POPULAR = 'popular',
  OLDEST = 'oldest',
}
export type Orientation = 'landscape' | 'portrait' | 'squarish';

export type PaginationParams = {
  perPage?: number;
  page?: number;
};
