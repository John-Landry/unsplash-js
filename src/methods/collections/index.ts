import { handleFeedResponse } from '../../helpers/feed';
import * as Query from '../../helpers/query';
import { createRequestHandler, createRequestHandlerOptional } from '../../helpers/request';
import { castResponse } from '../../helpers/response';
import { OrientationParam, PaginationParams } from '../../types/request';

type CollectionId = {
  collectionId: string;
};

const COLLECTIONS_PATH_PREFIX = '/collections';

export const getPhotos = {
  handleRequest: createRequestHandler(
    ({
      collectionId,
      orientation,
      ...paginationParams
    }: CollectionId & PaginationParams & OrientationParam) => ({
      pathname: `${COLLECTIONS_PATH_PREFIX}/${collectionId}/photos`,
      query: { ...Query.getFeedParams(paginationParams), orientation },
    }),
  ),
  handleResponse: handleFeedResponse<any>(),
};

export const get = {
  handleRequest: createRequestHandler(({ collectionId }: CollectionId) => ({
    pathname: `${COLLECTIONS_PATH_PREFIX}/${collectionId}`,
    query: {},
  })),
  handleResponse: castResponse<any>(),
};

export const list = {
  handleRequest: createRequestHandlerOptional(
    (paginationParams: Pick<PaginationParams, 'page' | 'perPage'> = {}) => ({
      pathname: COLLECTIONS_PATH_PREFIX,
      query: Query.getFeedParams(paginationParams),
    }),
  ),
  handleResponse: handleFeedResponse<any>(),
};

export const getRelated = {
  handleRequest: createRequestHandler(({ collectionId }: CollectionId) => ({
    pathname: `${COLLECTIONS_PATH_PREFIX}/${collectionId}/related`,
    query: {},
  })),
  handleResponse: castResponse<any>(),
};
