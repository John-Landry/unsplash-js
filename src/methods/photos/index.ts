import urlHelpers from 'url';
import { handleFeedResponse } from '../../helpers/feed';
import * as Query from '../../helpers/query';
import { createRequestParams } from '../../helpers/request';
import { castResponse } from '../../helpers/response';
import { isDefined } from '../../helpers/typescript';
import { OrientationParam, PaginationParams } from '../../types/request';

type PhotoId = {
  photoId: string;
};

const PHOTOS_PATH_PREFIX = '/photos';

export const list = {
  handleRequest: createRequestParams((feedParams: PaginationParams = {}) => ({
    pathname: PHOTOS_PATH_PREFIX,
    query: Query.getFeedParams(feedParams),
  })),
  handleResponse: handleFeedResponse<any>(),
};

export const get = {
  handleRequest: createRequestParams(({ photoId }: PhotoId) => ({
    pathname: `${PHOTOS_PATH_PREFIX}/${photoId}`,
  })),
  handleResponse: castResponse<any>(),
};

export const getStats = {
  handleRequest: createRequestParams(({ photoId }: PhotoId) => ({
    pathname: `${PHOTOS_PATH_PREFIX}/${photoId}/statistics`,
  })),
  handleResponse: castResponse<any>(),
};

const generateCacheBuster = () => new Date().getTime().toString();

export const getRandom = {
  handleRequest: createRequestParams(
    ({
      cacheBuster = generateCacheBuster(),
      collectionIds,
      ...queryParams
    }: {
      collectionIds?: string[];
      featured?: boolean;
      username?: string;
      query?: string;
      /**
       * Avoid response caching
       */
      cacheBuster?: string;
      count?: number;
    } & OrientationParam = {}) => ({
      pathname: `${PHOTOS_PATH_PREFIX}/random`,
      query: {
        ...queryParams,
        c: cacheBuster,
        ...Query.getCollections(collectionIds),
      },
    }),
  ),
  handleResponse: castResponse<any>(),
};

export const track = {
  handleRequest: createRequestParams(({ downloadLocation }: { downloadLocation: string }) => {
    const { pathname, query } = urlHelpers.parse(downloadLocation, true);

    if (!isDefined(pathname)) {
      throw new Error('Could not parse pathname from url.');
    }
    return { pathname, query };
  }),
  handleResponse: castResponse<any>(),
};
