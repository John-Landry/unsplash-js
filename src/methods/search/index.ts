import { createRequestParams } from '../../helpers/request';
import { APISearchFilters, ContentFilter, Language } from './types';
import * as Query from '../../helpers/query';
import { PaginationParams } from '../../types/request';
import { castResponse } from '../../helpers/response';

export type SearchParams = {
  query: string;
} & PaginationParams;

export type SearchParamsWithFilters = SearchParams &
  APISearchFilters & {
    lang?: Language;
    contentFilter?: ContentFilter;
    collectionIds?: string[];
  };

export const getPhotos = {
  handleRequest: ({
    query,
    page,
    perPage,
    collectionIds,
    lang,
    contentFilter,
    ...filters
  }: SearchParamsWithFilters) =>
    createRequestParams({
      pathname: '/search/photos',
      query: {
        query,
        content_filter: contentFilter,
        lang,
        ...Query.getFeedParams({ page, perPage }),
        ...Query.getCollections(collectionIds),
        ...filters,
      },
    }),
  handleResponse: castResponse<any>(),
};

export const getCollections = {
  handleRequest: ({ query, ...feedParams }: SearchParams) =>
    createRequestParams({
      pathname: '/search/collections',
      query: { query, ...Query.getFeedParams(feedParams) },
    }),
  handleResponse: castResponse<any>(),
};

export const getUsers = {
  handleRequest: ({ query, ...feedParams }: SearchParams) =>
    createRequestParams({
      pathname: '/search/users',
      query: { query, ...Query.getFeedParams(feedParams) },
    }),
  handleResponse: castResponse<any>(),
};
