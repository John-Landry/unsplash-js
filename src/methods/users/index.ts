import { createRequestParams } from '../../helpers/request';
import { Orientation, PaginationParams } from '../../types/request';
import * as Query from '../../helpers/query';

export const getProfile = ({ username }: { username: string }) =>
  createRequestParams({
    pathname: `/users/${username}`,
  });

export const getPhotos = ({
  username,
  stats,
  orientation,
  ...feedParams
}: {
  username: string;
  stats?: boolean;
  orientation: Orientation;
} & PaginationParams) =>
  createRequestParams({
    pathname: `/users/${username}/photos`,
    query: {
      ...Query.getFeedParams(feedParams),
      orientation,
      stats,
    },
  });

export const getLikes = ({
  username,
  orientation,
  ...feedParams
}: { username: string; orientation: Orientation } & PaginationParams) =>
  createRequestParams({
    pathname: `/users/${username}/likes`,
    query: {
      ...Query.getFeedParams(feedParams),
      orientation,
    },
  });

export const getCollections = ({
  username,
  ...feedParams
}: { username: string } & PaginationParams) =>
  createRequestParams({
    pathname: `/users/${username}/collections`,
    query: {
      ...Query.getFeedParams(feedParams),
    },
  });
