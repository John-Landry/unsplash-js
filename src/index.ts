import { initMakeRequest } from './helpers/request';
import * as photos from './methods/photos';
import * as search from './methods/search';
import * as users from './methods/users';
import * as collections from './methods/collections';
import { flow } from './helpers/fp';

export const Unsplash = flow(initMakeRequest, makeRequest => ({
  photos: {
    get: flow(photos.get, makeRequest),
    listFeed: flow(photos.listFeed, makeRequest),
    getStats: flow(photos.getStats, makeRequest),
    getRandom: flow(photos.getRandom, makeRequest),
    track: flow(photos.track, makeRequest),
  },
  users: {
    getPhotos: flow(users.getPhotos, makeRequest),
    getCollections: flow(users.getCollections, makeRequest),
    getLikes: flow(users.getLikes, makeRequest),
    get: flow(users.get, makeRequest),
  },
  search: {
    getCollections: flow(search.getCollections, makeRequest),
    getPhotos: flow(search.getPhotos, makeRequest),
    getUsers: flow(search.getUsers, makeRequest),
  },
  collections: {
    getPhotos: flow(collections.getPhotos, makeRequest),
    get: flow(collections.get, makeRequest),
    list: flow(collections.list, makeRequest),
    getRelated: flow(collections.getRelated, makeRequest),
  },
}));
