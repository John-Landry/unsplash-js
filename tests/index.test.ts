import { buildUrl, CompleteRequestParams } from '../src/helpers/request';
import * as collections from '../src/methods/collections';
import * as photos from '../src/methods/photos';
import * as search from '../src/methods/search';
import * as users from '../src/methods/users';
import { createApi } from '../src';
import { OrderBy } from '../src/types/request';

describe('buildUrl', () => {
  it('combines pathname, query and domain correctly', () => {
    const output = buildUrl({ pathname: '/foo/bar', query: { a: 1, b: 2 } })('https://example.com');
    expect(output).toEqual('https://example.com/foo/bar?a=1&b=2');
  });

  it('handles empty query correctly', () => {
    const output = buildUrl({ pathname: '/foo/bar' })('https://example.com');
    expect(output).toEqual('https://example.com/foo/bar');
  });

  it('handles undefined query properties correctly', () => {
    const output = buildUrl({ pathname: '/foo/bar', query: { a: 1, b: undefined } })(
      'https://example.com',
    );
    expect(output).toEqual('https://example.com/foo/bar?a=1');
  });

  it('handles query with only undefined properties correctly', () => {
    const output = buildUrl({ pathname: '/foo/bar', query: { b: undefined } })(
      'https://example.com',
    );
    expect(output).toEqual('https://example.com/foo/bar');
  });
});

const PHOTO_ID = 'abc123';
const USERNAME = 'usernametest';
const SEARCH_QUERY = 'cat';
const COLLECTION_ID = 'collection123';

type Section = keyof ReturnType<typeof createApi>;
const paramsTests: Record<Section, { [index: string]: CompleteRequestParams[] }> = {
  photos: {
    get: [photos.get.handleRequest({ photoId: PHOTO_ID })],
    list: [
      photos.list.handleRequest({ orderBy: OrderBy.LATEST, page: 4, perPage: 10 }),
      photos.list.handleRequest({ orderBy: OrderBy.LATEST }),
    ],
    getStats: [photos.getStats.handleRequest({ photoId: PHOTO_ID })],
    getRandom: [photos.getRandom.handleRequest({})],
    track: [
      photos.track.handleRequest({
        downloadLocation: 'https://api.unsplash.com/photos/foo123/download',
      }),
    ],
  },
  users: {
    getPhotos: [users.getPhotos.handleRequest({ username: USERNAME })],
    getCollections: [users.getCollections.handleRequest({ username: USERNAME })],
    getLikes: [users.getLikes.handleRequest({ username: USERNAME })],
    get: [users.get.handleRequest({ username: USERNAME })],
  },
  search: {
    getCollections: [search.getCollections.handleRequest({ query: SEARCH_QUERY })],
    getPhotos: [search.getPhotos.handleRequest({ query: SEARCH_QUERY })],
    getUsers: [search.getUsers.handleRequest({ query: SEARCH_QUERY })],
  },
  collections: {
    getPhotos: [collections.getPhotos.handleRequest({ collectionId: COLLECTION_ID })],
    get: [collections.get.handleRequest({ collectionId: COLLECTION_ID })],
    list: [collections.list.handleRequest({})],
    getRelated: [collections.getRelated.handleRequest({ collectionId: COLLECTION_ID })],
  },
};

describe('requestParams', () => {
  Object.entries(paramsTests).forEach(([sectionName, section]) => {
    describe(`${sectionName}`, () => {
      Object.entries(section).forEach(([methodName, methodTests]) => {
        describe(`${methodName}: `, () => {
          it('Correcty builds request arguments', () => {
            methodTests.forEach(output => expect(output).toMatchSnapshot());
          });
        });
      });
    });
  });
});
