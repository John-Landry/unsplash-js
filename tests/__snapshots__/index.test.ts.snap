// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requestParams collections get:  Correcty builds request arguments 1`] = `
Object {
  "headers": Object {},
  "pathname": "/collections/collection123",
  "queryToUse": Object {},
}
`;

exports[`requestParams collections getPhotos:  Correcty builds request arguments 1`] = `
Object {
  "headers": Object {},
  "pathname": "/collections/collection123/photos",
  "queryToUse": Object {},
}
`;

exports[`requestParams collections getRelated:  Correcty builds request arguments 1`] = `
Object {
  "headers": Object {},
  "pathname": "/collections/collection123/related",
  "queryToUse": Object {},
}
`;

exports[`requestParams collections list:  Correcty builds request arguments 1`] = `
Object {
  "headers": Object {},
  "pathname": "/collections",
  "queryToUse": Object {},
}
`;

exports[`requestParams photos get:  Correcty builds request arguments 1`] = `
Object {
  "headers": Object {},
  "pathname": "/photos/abc123",
  "queryToUse": Object {},
}
`;

exports[`requestParams photos getRandom:  Correcty builds request arguments 1`] = `
Object {
  "headers": Object {
    "cache-control": "no-cache",
  },
  "pathname": "/photos/random",
  "queryToUse": Object {},
}
`;

exports[`requestParams photos getStats:  Correcty builds request arguments 1`] = `
Object {
  "headers": Object {},
  "pathname": "/photos/abc123/statistics",
  "queryToUse": Object {},
}
`;

exports[`requestParams photos list:  Correcty builds request arguments 1`] = `
Object {
  "headers": Object {},
  "pathname": "/photos",
  "queryToUse": Object {
    "order_by": "latest",
    "page": 4,
    "per_page": 10,
  },
}
`;

exports[`requestParams photos list:  Correcty builds request arguments 2`] = `
Object {
  "headers": Object {},
  "pathname": "/photos",
  "queryToUse": Object {
    "order_by": "latest",
  },
}
`;

exports[`requestParams photos track:  Correcty builds request arguments 1`] = `
Object {
  "headers": Object {},
  "pathname": "/photos/foo123/download",
  "queryToUse": Object {},
}
`;

exports[`requestParams search getCollections:  Correcty builds request arguments 1`] = `
Object {
  "headers": Object {},
  "pathname": "/search/collections",
  "queryToUse": Object {
    "query": "cat",
  },
}
`;

exports[`requestParams search getPhotos:  Correcty builds request arguments 1`] = `
Object {
  "headers": Object {},
  "pathname": "/search/photos",
  "queryToUse": Object {
    "query": "cat",
  },
}
`;

exports[`requestParams search getUsers:  Correcty builds request arguments 1`] = `
Object {
  "headers": Object {},
  "pathname": "/search/users",
  "queryToUse": Object {
    "query": "cat",
  },
}
`;

exports[`requestParams users get:  Correcty builds request arguments 1`] = `
Object {
  "headers": Object {},
  "pathname": "/users/usernametest",
  "queryToUse": Object {},
}
`;

exports[`requestParams users getCollections:  Correcty builds request arguments 1`] = `
Object {
  "headers": Object {},
  "pathname": "/users/usernametest/collections",
  "queryToUse": Object {},
}
`;

exports[`requestParams users getLikes:  Correcty builds request arguments 1`] = `
Object {
  "headers": Object {},
  "pathname": "/users/usernametest/likes",
  "queryToUse": Object {},
}
`;

exports[`requestParams users getPhotos:  Correcty builds request arguments 1`] = `
Object {
  "headers": Object {},
  "pathname": "/users/usernametest/photos",
  "queryToUse": Object {},
}
`;
